data-prepper-plugins/s3-source/src/main/java/com/amazon/dataprepper/plugins/source/codec/CSVCodec.java

```diff
@@ -0,0 +1,145 @@
+
+    private void parseBufferedReader(final BufferedReader reader, final Consumer<Record<Event>> eventConsumer) throws IOException {
+        final CsvMapper mapper = createCsvMapper();
+        final CsvSchema schema;
+        if (config.isDetectHeader()) {
+            schema = createAutodetectHeaderCsvSchema();
+        }
+        else {
+            final int numberColsFirstLine = getNumberOfColsByMarkingBeginningOfInputStreamAndResettingReaderAfter(reader);
+            schema = createCsvSchemaFromConfig(numberColsFirstLine);
+        }
+
+        MappingIterator<Map<String, String>> parsingIterator = mapper.readerFor(Map.class).with(schema).readValues(reader);
+        try {
+            while (parsingIterator.hasNextValue()) {
+                readCSVLine(parsingIterator, eventConsumer);
+            }
+        } catch (Exception jsonExceptionOnHasNextLine) {
+            LOG.error("An Exception occurred while determining if file has next line ", jsonExceptionOnHasNextLine);
+        }
+    }
+
+    private int getNumberOfColsByMarkingBeginningOfInputStreamAndResettingReaderAfter(BufferedReader reader) throws IOException {
+        final int defaultBufferSize = 8192; // this number doesn't affect even a thousand column header — it's sufficiently large.
+        reader.mark(defaultBufferSize); // calling reader.readLine() will consume the first line, so mark initial location to reset after
+        final int firstLineNumCols = extractNumberOfColsFromFirstLine(reader.readLine());
+        reader.reset(); // move reader pointer back to beginning of file in order to reread first line
+        return firstLineNumCols;
+    }
+
+    private void readCSVLine(final MappingIterator<Map<String, String>> parsingIterator, final Consumer<Record<Event>> eventConsumer) {
+        try {
+            final Map<String, String> parsedLine = parsingIterator.nextValue();
+
+            final Event event = JacksonLog.builder()
+                    .withData(parsedLine)
+                    .build();
+            eventConsumer.accept(new Record<>(event));
+        } catch (final CsvReadException csvException) {
+            LOG.error("Invalid CSV row, skipping this line. This typically means the row has too many lines. Consider using the CSV " +
+                    "Processor if there might be inconsistencies in the number of columns because it is more flexible. Error ",
+                    csvException);
+        } catch (JsonParseException jsonException) {
+            LOG.error("A JsonParseException occurred on a row of the CSV file, skipping line. This typically means a quote character was " +
+                    "not properly closed. Error ", jsonException);
+        } catch (final Exception e) {
+            LOG.error("An Exception occurred while reading a row of the CSV file. Error ", e);
+        }
+    }
+
+    private int extractNumberOfColsFromFirstLine(final String firstLine) {
+        int numberOfSeperators = 0;
+        int charPointer = 0;
```

Comment:
Is there any reason this is not in the `for` statement?

Reply:
There's no reason, oversight on my part. Thanks for catching it!